### ISSUE TITLE
Can't close Netty channel with NettyTcpClient

### ISSUE THREAD

**Issue by rstoyanchev** [reactor/reactor#510]
Completing the `Publisher<Void>` returned when a `ReactorChannelHandler` is applied doesn't close the underlying Netty channel. From what I can see in [NettyChannelHandlerBridge](https://github.com/reactor/reactor/blob/9159926d2c75d6b7808c75ab51deacbc91499b51/reactor-net/src/main/java/reactor/io/net/impl/netty/NettyChannelHandlerBridge.java#L122-L126) the close is protected with an if condition that ensures there isn't a subscriber to the channel. The only way for channelSubscription to become null is after [the channel is closed](https://github.com/reactor/reactor/blob/9159926d2c75d6b7808c75ab51deacbc91499b51/reactor-net/src/main/java/reactor/io/net/impl/netty/NettyChannelHandlerBridge.java#L156-L159). So as far as I can see there is no way to close the connection.

This was reported under [SPR-13290](https://jira.spring.io/browse/SPR-13290) where if a heartbeat from the broker comes late (or never), we close the connection. In turn that should trigger a reconnect through the [ReconnectingChannelListener](https://github.com/reactor/reactor/blob/9159926d2c75d6b7808c75ab51deacbc91499b51/reactor-net/src/main/java/reactor/io/net/impl/netty/tcp/NettyTcpClient.java#L274) but it doesn't since the channel is not closed.

I'll include some sample code.

**Comment by rstoyanchev** [comment]
Bring up a RabbitMQ instance with the [STOMP adapter](https://www.rabbitmq.com/stomp.html) installed up and running locally with default settings.

Then run the code below, which should connect to the broker and start receiving heartbeats every 5 seconds. Then stop the broker (`sudo rabbitmqctl stop`), see the reconnect attempts printed out, then start again (`sudo rabbitmq-server -detached`), see the connection re-established.

Then run the code again with the commented out line that closes the connection, commented in. The connection doesn't get actually closed, we continue to receive heartbeats.

```` java
import java.io.IOException;

import org.springframework.messaging.Message;
import org.springframework.messaging.simp.stomp.Reactor2StompCodec;
import org.springframework.messaging.simp.stomp.StompCommand;
import org.springframework.messaging.simp.stomp.StompDecoder;
import org.springframework.messaging.simp.stomp.StompEncoder;
import org.springframework.messaging.simp.stomp.StompHeaderAccessor;
import org.springframework.messaging.support.MessageBuilder;
import org.springframework.messaging.support.MessageHeaderAccessor;
import org.springframework.messaging.tcp.ReconnectStrategy;
import org.springframework.messaging.tcp.TcpConnection;
import org.springframework.messaging.tcp.TcpConnectionHandler;
import org.springframework.messaging.tcp.TcpOperations;
import org.springframework.messaging.tcp.reactor.Reactor2TcpClient;

public class Test {

	public static void main(String[] args) throws IOException {

		StompDecoder decoder = new StompDecoder();
		Reactor2StompCodec codec = new Reactor2StompCodec(new StompEncoder(), decoder);
		TcpOperations<byte[]> tcpClient = new Reactor2TcpClient<byte[]>("127.0.0.1", 61613, codec);

		ReconnectStrategy reconnect = new ReconnectStrategy() {
			@Override
			public Long getTimeToNextAttempt(int attemptCount) {
				System.out.println("Reconnect attempt: " + attemptCount);
				return 1000L;
			}
		};

		tcpClient.connect(new TestTcpConnectionHandler(), reconnect);

		System.in.read();
	}

	private static class TestTcpConnectionHandler implements TcpConnectionHandler<byte[]> {

		TcpConnection<?> connection;

		@Override
		public void afterConnected(TcpConnection<byte[]> connection) {
			System.out.println("TCP connection established.");
			this.connection = connection;

			StompHeaderAccessor accessor = StompHeaderAccessor.create(StompCommand.CONNECT);
			accessor.setAcceptVersion("1.1,1.2");
			accessor.setLogin("guest");
			accessor.setPasscode("guest");
			accessor.setHeartbeat(0, 5000);
			connection.send(MessageBuilder.createMessage(new byte[0], accessor.getMessageHeaders()));
		}

		@Override
		public void afterConnectFailure(Throwable ex) {
			System.out.println("afterConnectFailure");
			ex.printStackTrace();
		}

		@Override
		public void handleMessage(Message<byte[]> message) {
			StompHeaderAccessor accessor = MessageHeaderAccessor.getAccessor(message, StompHeaderAccessor.class);
			System.out.println(accessor.getDetailedLogMessage(message.getPayload()));

			if (StompCommand.CONNECTED.equals(accessor.getCommand())) {
				System.out.println("Closing connection (from our end).");
				// this.connection.close();
			}
		}

		@Override
		public void handleFailure(Throwable ex) {
			System.out.println("Failure");
			ex.printStackTrace();
		}

		@Override
		public void afterConnectionClosed() {
			System.out.println("Connection closed");
		}
	}

}
````

**Comment by rstoyanchev** [comment]
Note that the issue exists in both Reactor 2.x and 1.x. It's been around for some time.

**Comment by rstoyanchev** [comment]
I checked to see if this can be done by cancelling the channelSubscription but as far as I can tell [that's commented out](https://github.com/reactor/reactor/blob/9159926d2c75d6b7808c75ab51deacbc91499b51/reactor-net/src/main/java/reactor/io/net/impl/netty/NettyChannelHandlerBridge.java#L77-L93) and I don't see any other code in NettyChannelHandlerBridge that can close the connection.

**Comment by dangduc** [comment]
We've been experiencing this issue in production for months. Are there any solid plans to resolve this issue?

**Comment by smaldini** [comment]
now available in 2.0.6 snapshot on repo.spring.io

**Comment by dangduc** [comment]
Thanks for working on this. This appears to fix our production issue. Thank you very much.
